### PR TITLE
Feat/change roles exist

### PR DESCRIPTION
**Pull request by Marcellofabbri**
**TIS21-1629**: Update bulk person create to validate and set the role correctly

Third approach. Rather than acting on TCS, the actual rolesExist endpoint in REFERENCE (RoleResource) gets changed.
It's supposed to work in a non case-sensitive way, so that roles (the codes of the roles) that exist (but with different casing) are returned in a Map as before. What is returned is a Map<String, String> where the key is the code that is passed in, and the value is the role as found in the database (correctly spelt), or an empty string if there are no matches.

This approach aims at disrupting TCS as least as possible. It'll still need to have PersonValidator changed in order to set the correctly spelt role code as the `PersonDto.role`, as per below:
```
TCS - PersonValidator.java

private List<FieldError> checkRole(PersonDTO personDto) {
    List<FieldError> fieldErrors = new ArrayList<>();

    String roleMultiValue = personDto.getRole();

    if (!StringUtils.isEmpty(roleMultiValue)) {
      // Bulk upload uses a ";" separator, account for both that and the default ",".
      List<String> roles = Arrays.stream(roleMultiValue.split("[;,]"))
          .map(String::trim)
          .collect(Collectors.toList());

      Map<String, String> rolesExist = referenceService.rolesExist(roles, true);
      List<String> correctedRoles = rolesExist.entrySet().stream()
          .map(entry -> entry.getValue().isEmpty() ? entry.getKey() : entry.getValue())
          .collect(Collectors.toList());

      personDto.setRole(String.join(",", correctedRoles));

      for (Entry<String, String> roleExists : rolesExist.entrySet()) {
        if (roleExists.getValue().isEmpty()) {
          fieldErrors.add(new FieldError(PERSON_DTO_NAME, "role",
              String.format("Role '%s' did not match a reference value.", roleExists.getKey())));
        }
      }
    }
```
Here's the TCS PR: [TCS PR 756](https://github.com/Health-Education-England/TIS-TCS/pull/756)

I tried to see if the `rolesExist` endpoint was being called from anywhere else where its slight change of behavior could determine undesired results, but I couldn't find any examples.
As far as I could tell, it seems to work fine locally when both tweaked versions of REFERENCE and TCS are running. Wrong codes get rejected, existing codes get accepted, codes that exist but with different casing get accepted and written with correct casing.

### CODE REVIEW

**Comment by Judge40** [comment]
I'm struggling to follow some of the code, which might suggest a problem in complexity, so I'm going to pull it down locally to step through.

While I'm doing that though, I'm not sure I'm a fan of changing this endpoint in this way. Although it isn't used anywhere else it still feels odd to answer the question "Does this exist?" with a map where the key doesn't match the original input. It doesn't really simplify the work of the caller either, since they still have to do case insenstive matching on the returned values - all it saves them is the hassle of getting all the current roles.

An alternative might be to add a new/replacement "autocorrect" or "findClosestMatch" type endpoint with a return of `Map<String,String>` where the key is the original input and the value is the correctly cased value. It gives us a bit more scope to add more complex autocorrect/matching if needed in the future and gives the caller more context about what we've done.

**Comment by Marcellofabbri** [comment]
If the endpoint isn't used anywhere else, can't we just rename this one to "autocorrect casing" and make it return the Map<String,String> instead of the Map<String, Boolean>?

**Comment by Marcellofabbri** [comment]
I've implemented the use of Map<String,String> and made it return the pair ("originalCode", "") in case the role code not found.
I haven't changed the names of the methods and endpoint url, so these still need to be addressed if we decide not to call this "exist"

**Comment by ReubenRobertsHEE** [comment]
I wonder how we can progress this? I've taken a look and it seems to work fine as it stands, though I would agree it is not exactly what you might assume you'd get from the endpoint - but an update to the javadoc should clarify that. 

There was some concern at stand-up about consistency across the reference services, I think @hihilary you said they are case-sensitive so if we change this one to insensitive then things might become inconsistent? Does that refer to this work you were doing on 26 Sep in TCS https://github.com/Health-Education-England/TIS-TCS/commit/b39ab803be351c1f77e7bbc26507e8a26c2a7b0d ? I am quite big on consistency as well, but would you be able to clarify where the issue is? As far as I can see, in the reference service there are plenty of other endpoints that are case-insensitive.

e.g. for roles:

`getRolesIn` - case-insensitive e.g. https://apps.tis.nhs.uk/reference/api/roles/in/ArCP%20Lay%20Member == https://apps.tis.nhs.uk/reference/api/roles/in/ArCP%20LAY%20Member

also for sites:

`findSitesIn` - case insensitive, e.g. https://apps.tis.nhs.uk/reference/api/sites/in/M86003 == https://apps.tis.nhs.uk/reference/api/sites/in/m86003

`findSitesByName` - case insensitive, e.g. http://local.tis.com/reference/https://apps.tis.nhs.uk/reference/api/sites?columnFilters=%7B+%22siteKnownAs%22%3A+%5B%22DEnTal%20Surgery%20(V02974)%22%5D%2C%22status%22%3A%5B%22CURRENT%22%5D%7D == https://apps.tis.nhs.uk/reference/api/sites?columnFilters=%7B+%22siteKnownAs%22%3A+%5B%22Dental%20Surgery%20(V02974)%22%5D%2C%22status%22%3A%5B%22CURRENT%22%5D%7D

I haven't looked any further since that might start feeling like an audit, which is probably something that needs a ticket :)

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (ReubenRobertsHEE) [inline reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:220]: I think we need the javadoc updated here :)

**Comment by ReubenRobertsHEE** [comment]
If we are sure that the original `/roles/exists/` endpoint is not needed or used anywhere then I guess we can scrap it, but otherwise I'd suggest putting the new work into a new endpoint as Marcello and Andy suggest - I'd vote for `/roles/matches/` .

**Comment by Marcellofabbri** [comment]
All good point @ReubenRobertsHEE, I've just pushed a commit that tries to align the nomenclature with 'match' rather than 'exist', I hope I caught all the references to "roles existing" and turned them into "roles being matched", if it makes sense

**Comment by hihilary** [comment]
@ReubenRobertsHEE I tried some more endpoint, and think I was wrong, queries in mysql are case insensitive by default. Sorry I brought confusion, should have digged a little more.

**Comment by Marcellofabbri** [comment]
I've added the back endpoint back in (on top of the new one), so now it has both the new one (`/roles/matches/`) and old one (`roles/exists/`). If deployed, it shouldn't break TCS, which will still use the old endpoint until we merge its PR

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/service/impl/SpecificationFactory.java
Comment (hihilary) [inline reference-service/src/main/java/com/transformuk/hee/tis/reference/service/service/impl/SpecificationFactory.java:112]: Comparing to the original `in` method, this line is the difference. But the `in` method already deals with case insensitivity, so `inIgnoreCase` is not required.

File: reference-client/src/test/java/com/transformuk/hee/tis/reference/client/impl/ReferenceServiceImplTest.java
Comment (ReubenRobertsHEE) [inline reference-client/src/test/java/com/transformuk/hee/tis/reference/client/impl/ReferenceServiceImplTest.java:1269]: Would you mind restoring the old tests for exists() and implementing new tests for matches() - that way the code coverage won't decrease.

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (ReubenRobertsHEE) [inline reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:248]: I'd say 'EXISTS' should be replaced with 'POST' here. I've not seen the term 'EXISTS' before in a javadoc - it sort of makes sense to me in the rolesExists() definition, but all the other javadocs here simply include the HTTP method type, which might be better to stick with in this instance.

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (ReubenRobertsHEE) [inline reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:263]: Its ok here because its debug logging only, but in general this can be vulnerable to log injection - we had to do some clean-ups on some of the recent projects we open-sourced to guard against this where the log level was 'info' and URL parameters were being logged as-is.

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (ReubenRobertsHEE) [reply reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:278]: Here's a simpler example:
`curl -X POST http://local.tis.com/reference/api/roles/matches/ -H 'cache-control: no-cache' -H "Content-Type: application/json" -d '["ARCP/Recruitment Assessor","ARCP/Recruitment Assessor"]'`

File: reference-client/src/test/java/com/transformuk/hee/tis/reference/client/impl/ReferenceServiceImplTest.java
Comment (Marcellofabbri) [reply reference-client/src/test/java/com/transformuk/hee/tis/reference/client/impl/ReferenceServiceImplTest.java:1323]: I've restored them for both ReferenceServiceImpl and RoleResource. Originally I was going to simply change the old endpoint into the new one, and that's why I simply amended the old tests to fit the new endpoint.  Then I put the old endpoint back in because otherwise TCS would break until we merge its PR to make it use the new endpoint.
We may end up deleting these old tests though if we go back to REFERENCE later to delete the old endpoint once TCS passed to the new one.

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (Marcellofabbri) [reply reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java]: True

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (Marcellofabbri) [reply reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:263]: So I'll leave it in?

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (Marcellofabbri) [reply reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:278]: Well spotted. This appears to have always been a vulnerability all along since the old endpoint also returns a Map, and by passing two duplicate roles into a Person row it currently fails when it tries to put two duplicate keys in the Map (the error on the spreadsheet simply reads 'Internal Server Error').

I don't know whether it's better to:

1. Prevent duplicates from TCS, right before it calls the endpoint, so that two duplicates are never passed in
or
2. Make the endpoint duplicate-safe, so that it returns a specific exception that TCS will then have to handle

The first approach leaves the endpoint a bit vulnerable and leaves it up to the developers to use it sensibly (i.e. put checks in place in TCS before it passes duplicates to it).
The second approach makes the endpoint safer, so that there are less chances for anybody to misuse it, even in possible future developments.

I feel like the second one is the best approach, even if it makes it a bit harder to then capture that failure in TCS to handle it correctly when returning the PersonDto.

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (ReubenRobertsHEE) [reply reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:263]: Yeah, its fine - it was just a comment for interest :)

File: reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java
Comment (ReubenRobertsHEE) [reply reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:278]: I wonder if it should simply return a map containing the distinct values that were passed in, and not throw an exception? To me, that would be reasonable. If the calling process can't handle the fact that the map it's getting back is not the same length as the list it passed in (if there were duplicates in the list), then that would seem to me to be a problem for the calling process to sort out itself.

Comment (ReubenRobertsHEE) [reply reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:278]: So basically, (2) above, except not throwing an exception.

Comment (ReubenRobertsHEE) [reply reference-service/src/main/java/com/transformuk/hee/tis/reference/service/api/RoleResource.java:280]: Sorry for the late response, I don't seem to be getting alerts for these updates.

Ah, apologies - I should have been clearer when I made my suggestion. I actually thought this commit was perfect: https://github.com/Health-Education-England/TIS-REFERENCE/pull/202/commits/76b5e6f916fe8236a9166e865aa3d0a8aad6b7ea!

To me its slightly odd to filter out case-insensitive duplicates in the codes submitted, but to use the first one (with its casing intact) as the 'master' that is returned.

However, this is maybe getting into hair-splitting territory for some quite far-out edge cases. I'm happy to proceed with either version. If you want to go back to the previous commit let me know, otherwise I'll approve after stand-up.

File: reference-client/src/test/java/com/transformuk/hee/tis/reference/client/impl/ReferenceServiceImplTest.java
Comment (ReubenRobertsHEE) [inline reference-client/src/test/java/com/transformuk/hee/tis/reference/client/impl/ReferenceServiceImplTest.java:1342]: Last thing I think (sorry Marcello!) - 'exists' should be 'matches' here and in the test below.

Comment (Marcellofabbri) [reply reference-client/src/test/java/com/transformuk/hee/tis/reference/client/impl/ReferenceServiceImplTest.java]: ahah it's fine, there were loads of 'exists' to weed out and swap with 'matches', I knew I must have left some behind. Well spotted!

**Review by ReubenRobertsHEE — APPROVED** [review]

**Pull request merged**
